src/util/finite_field.h

```diff
@@ -0,0 +1,195 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Alex Ozdemir
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * A finite-field element, implemented as a wrapper around Integer.
+ *
+ * TODOs:
+ * * consider montgomery form
+ * * extend to non-prime fields
+ */
+
+#include "cvc5_public.h"
+
+#ifndef CVC5__FINITE_FIELD_H
+#define CVC5__FINITE_FIELD_H
+
+#include <iosfwd>
+#include <iostream>
+
+#include "base/check.h"
+#include "base/exception.h"
+#include "util/integer.h"
+
+namespace cvc5::internal {
+
+class FiniteField
+{
+ public:
+  FiniteField(const Integer& val, const Integer& size)
+      // for the normalization
+      : d_size(size), d_value(val.floorDivideRemainder(size))
+  {
+    // we only support prime fields right now
+    CheckArgument(size.isProbablePrime(), size);
+  }
+
+  /**
+   * Construct the zero in this field
+   */
+  FiniteField(const Integer& size) : d_size(size), d_value(0)
+  {
+    // we only support prime fields right now
+    CheckArgument(size.isProbablePrime(), size);
+  }
+
+  ~FiniteField() {}
+
+  FiniteField& operator=(const FiniteField& x)
+  {
+    if (this == &x) return *this;
+    d_size = x.d_size;
+    d_value = x.d_value;
+    return *this;
+  }
+
+  /* Get value. */
+  const Integer& getValue() const;
+
+  /* Get field size. */
+  const Integer& getFieldSize() const;
+
+  /* Return value. */
+  Integer toInteger() const;
+  /* Return Integer of smallest absolute value that represents this.
+   *
+   * For fields of odd size, there is always a unique representative of
+   * smallest absolute value.
+   *
+   * For GF(2), the multiplicative identity is represented as "1", not "-1".
+   * */
+  Integer toSignedInteger() const;
+  /* Return (binary) string representation. */
+  std::string toString(unsigned int base = 2) const;
+
+  /* Return hash value. */
+  size_t hash() const;
+
+  /* -----------------------------------------------------------------------
+   ** Operators
+   * ----------------------------------------------------------------------- */
+
+  /* (Dis)Equality --------------------------------------------------------- */
+
+  /* Return true if this is equal to 'y'. */
+  bool operator==(const FiniteField& y) const;
```

Comment:
For binary operators, it is usually better to implement them as non-member functions [to maintain symmetry](https://en.cppreference.com/w/cpp/language/operators).

Reply:
Good point. We may want to fix BV at some point too.

Reply:
https://github.com/cvc5/cvc5/pull/9048